[Woo POS] Allow item list switch between Products and Coupons

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #15346 
<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

<!-- 
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
🛠 Need an AI review?  
Add a comment: `@coderabbitai review`
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
-->

## Description
This PR hooks the item provider with the UI, allowing the item list to either display a list of Products/Variations or Coupons on tapping a button, so it provides a basic interface to access those coupons and add them to the cart (future PR) from the dashboard.

https://github.com/user-attachments/assets/0d5f7aab-83a2-4c5e-9290-b005e49d3773

## Changes
- Switched the feature flag off to not bother development in trunk, since now we have some dummy UI
- Created a `POSItemType`
- Make the items controller to switch between fetching products/variations or coupons from storage based on the `POSItemType` state.
- Added dummy buttons to the dashboard, which only render under feature flag, so different types can be loaded as a list.
- Updated the coupons returning type from `[POSItem]` to `PagedItems<POSItem>` so these can be listed.

## Testing
- Set `enableCouponsInPointOfSale` flag to `true`
- Using a store with coupons enabled in WC settings, assure that we have coupons in local storage by going to Menu > Coupons
- Run the app, load POS, observe that products are loaded normally and variations still work as expected. 
- Tap the "Coupons" button and observe how these are shown in a list. Trying to add to cart will just display a debug log to the console.
- Switching back to products should load product list normally after a second or two, the loading state and transitions haven't been implemented yet.

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
Return coupons as PagedItems, hook to item list

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
Make providePointOfSaleCoupons throwable

Fetching from storage can fail, also we can reuse the error handling from items controller

iamgabrielma committed:
switch coupons feature flag to false

Now that we have dummy UI updates we should keep the flag off in trunk

iamgabrielma committed:
Make CouponCardView

iamgabrielma committed:
Remove unnecessary state. Collapse func duplication

iamgabrielma committed:
lint

iamgabrielma committed:
Invert fetch logic, make products first case

iamgabrielma committed:
lint

iamgabrielma commented:
@coderabbitai review

iamgabrielma commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -92,6 +92,11 @@ protocol PointOfSaleAggregateModelProtocol {
 @available(iOS 17.0, *)
 extension PointOfSaleAggregateModel {
+    @MainActor
+    func toggleItemType() async {
```
Would you say we should be explicitly of what type we pass through? I went with "toggle" for simplicity for now, but I'm always wary of ending in the wrong toggle state if we don't pass the property through here 🤔

staskus replied:
> Would you say we should be explicitly of what type we pass through?

Yes 👍 We already now we'll have at least a third option - customers, so the choice can be represented with enum.

iamgabrielma committed:
make test target compile

staskus commented:
@iamgabrielma This got away from my radar. I'll review this PR tomorrow!

staskus commented on WooCommerce/Classes/POS/Controllers/PointOfSaleItemsController.swift:
```diff
@@ -6,6 +6,11 @@ import enum Yosemite.PointOfSaleItemServiceError
 import struct Yosemite.POSVariableParentProduct
 import class Yosemite.Store
 
+enum POSItemType {
```
Based on the discussion p1741916795290219-slack-C070SJRA8DP, lets leave `PointOfSaleItemsController` and service for products, and have a separate one for coupons.

Since Coupons and Products are very different it makes sense to keep the logic separated. The controller doesn't need to know the current selection made in the UI.

staskus commented on WooCommerce/Classes/POS/Controllers/PointOfSaleItemsController.swift:
```diff
@@ -25,6 +32,7 @@ protocol PointOfSaleItemsControllerProtocol {
     private let paginationTracker: AsyncPaginationTracker
     private var childPaginationTrackers: [POSItem: AsyncPaginationTracker] = [:]
     private let itemProvider: PointOfSaleItemServiceProtocol
+    private var itemType: POSItemType = .products
```
Based on a discussion p1742206171389629/1741916795.290219-slack-C070SJRA8DP, we could also consider how `ItemListState` could be used to represent the Products | Coupons data structure. 

Maybe something like this? 🤔 

<img width="715" alt="image" src="https://github.com/user-attachments/assets/b27b59e5-c079-4286-9e93-c77471311d7f" />

The current selection could be tracked on a state or separately.

staskus reviewed (commented):
Thanks!

I think this is important first step so I'll delay approving it since we could structure it differently based on our discussions. 👍

iamgabrielma commented:
Sounds good! I've spent some time today experimenting with a new approach from what I've gathered from our conversation in slack:  https://github.com/woocommerce/woocommerce-ios/compare/trunk...experiment/pos-coupon-service 

This is not ready for review yet, but just a draft so we can sync. Let me know what do you think in general:

- We have a separate `PointOfSaleCouponsController`
- We have a separate `PointOfSaleCouponService`
- We have a `currentController` in the aggregate model, which can be either product or coupon one, both have their respective items loaded in memory, from the UI we just switch what we load based on `ItemListState`
- `ItemListState` and a separation of ItemType by product and coupons
- `ItemsStackState` becomes more complex, as needs to handle both types, so we may need to modify this structure.

There's also a blocker which I haven't figured out yet as I ran out of time: The app crashes when we switch between products/coupons, I'm guessing because miss-referencing storage when we switch services, I'll continue tomorrow with this.

staskus commented:
Thanks, @iamgabrielma!

> We have a separate PointOfSaleCouponsController
> We have a separate PointOfSaleCouponService

Sounds good! If you prefer, you could make PR with just of creation of the service and controller, which could be reviewed while we figure out the rest.

> We have a currentController in the aggregate model, which can be either product or coupon one, both have their respective items loaded in memory, from the UI we just switch what we load based on ItemListState

> We have a currentController in the aggregate model, which can be either product or coupon one, both have their respective items loaded in memory, from the UI we just switch what we load based on ItemListState

Another option would be to have an `enum` that represents the current selection which then could be used to select an appropriate controller. However, it doesn't change the approach, and `currentController` may be the best option. Likely `currentController` could be computer var that will return a controller based on `itemType`, which would hold the state of the current selection. This selection may be important to represent current selection in the UI.

> ItemListState and a separation of ItemType by product and coupons
> ItemsStackState becomes more complex, as needs to handle both types, so we may need to modify this structure.

I can imagine. Without actually doing it and seeing how it all works it's hard to advice something tangible. I'm not 100% sure what's the most comfortable way to represent it all.

staskus commented:
@iamgabrielma 

> We have a separate PointOfSaleCouponsController
> We have a separate PointOfSaleCouponService

By looking at the code, they will need to depend on a different protocol  than `PointOfSaleItemServiceProtocol`. We don't need knowledge of products within coupons and vice-versa. I guess it's all still WIP.

staskus commented:
> There's also a blocker which I haven't figured out yet as I ran out of time: The app crashes when we switch between products/coupons, I'm guessing because miss-referencing storage when we switch services, I'll continue tomorrow with this.

I think some storage code only expects to be called from the main actor. It's enough to make `PointOfSaleCouponService`  `providePointOfSaleCoupons` and `public func providePointOfSaleItems(pageNumber: Int) async throws -> PagedItems<POSItem>` to `@MainActor` to avoid the crash 👍 

I pushed the fix into the branch.

staskus commented:
@iamgabrielma 

Overall, I think the approach is good 👍 

Since we switch between `currentController`, we could maybe even avoid creating `SectionState` in the first PR, and just return the old `ItemListState`. This PR could focus on Service + Controller + Basic Switching.

Then we could move on to the required changes to the item state to support our needs. I don't want to put too many changes in one place. We may not even need `SectionState`, since `currentSectionState` simply returns `ItemListState.SectionState` which is of the same type.

WDYT?

iamgabrielma committed:
Merge branch 'trunk' into task/15346-pos-allow-switch-product-vs-coupon-list

iamgabrielma committed:
Make PointOfSaleCouponsController

iamgabrielma committed:
Create PointOfSaleCouponService

iamgabrielma committed:
Delete coupon-specific logic from itemservice

iamgabrielma committed:
Inject couponsController into POS entry point

iamgabrielma committed:
Hook UI item type selection with switch in aggregate model

iamgabrielma committed:
Inject couponsController in aggregate model and update item view state to use current controller

iamgabrielma committed:
Hook switching current viewstate to different controller

iamgabrielma committed:
lint

iamgabrielma committed:
Revert dummy CouponCardView

It's unnecessary in this PR as we do not handle rendering from the new coupon service, neither adding to cart

iamgabrielma commented:
That sounds good, thanks for the tips @staskus ! This PR is updated, and ready for review. I've limited the scope to:

- Created a new `PointOfSaleCouponsController`
- Created a new `PointOfSaleCouponService`
- Handle switch between different controllers in the aggregate model based on product type when we tap the button.
- Kept the `SectionState` as it is
- I've removed the coupon rendering and "add to cart" to handle later, but we can see in the console that the right path was taken when switching controllers, for now we'll just get an empty list of coupons. Eg:

```
"🍍 CouponsController::refreshItems called"
```

https://github.com/user-attachments/assets/adb997c1-e8e5-45cb-9934-d3a814260870

iamgabrielma commented on Yosemite/Yosemite/PointOfSale/PointOfSaleCouponService.swift:
```diff
@@ -0,0 +1,81 @@
+        self.siteID = siteID
+        self.currencyFormatter = CurrencyFormatter(currencySettings: currencySettings)
+        self.productsRemote = ProductsRemote(network: network)
+        self.variationRemote = ProductVariationsRemote(network: network)
```
I've kept this as it is for now, but we will need to split the `PointOfSaleItemServiceProtocol`, as coupons do not need to know about some of these dependencies.

staskus replied:
No need for splitting.

A separate `PointOfSaleCouponServiceProtocol ` is enough. We don't need the ability to interchangeably use `PointOfSaleCouponService` and `PointOfSaleItemService`. So we can safely remove all those methods and have:

```
public protocol PointOfSaleCouponServiceProtocol {
    func providePointOfSaleCoupons(pageNumber: Int) async throws -> PagedItems<POSItem>
}
```

iamgabrielma replied:
Gotcha! I'll take a look into this in the next PR as I start to implement those.

iamgabrielma committed:
cleanup unused

staskus commented on WooCommerce/Classes/POS/Controllers/PointOfSaleCouponsController.swift:
```diff
@@ -0,0 +1,36 @@
+
+@available(iOS 17.0, *)
+@Observable final class PointOfSaleCouponsController: PointOfSaleItemsControllerProtocol {
+    let itemType: ItemType = .coupons
```
Is `ItemType` supposed to be used in Controllers?

iamgabrielma replied:
It doesn't seem necessary, nope. Removed! https://github.com/woocommerce/woocommerce-ios/pull/15366/commits/249f75fe038b77b6f026966e1e01f7d758cc1368

staskus commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -120,6 +131,7 @@ private extension POSItem {
         case .variableParentProduct:
             return nil
         case .coupon:
+            debugPrint("Not implemented. TODO: Make POSCoupon POSOrderable")
```
Note: [May not need to be POSOrderable](https://github.com/woocommerce/woocommerce-ios/pull/15407), since for coupons it's enough to pass code to the API.

staskus reviewed (approved):
🚢 Looks good! Lets goooo

iamgabrielma committed:
Merge branch 'trunk' into task/15346-pos-allow-switch-product-vs-coupon-list

# Conflicts:
#	WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift
#	WooCommerce/Classes/POS/Presentation/ItemListView.swift
#	Yosemite/Yosemite/PointOfSale/PointOfSaleItemService.swift

iamgabrielma committed:
ammend merge conflict

iamgabrielma committed:
remove unnecessary prop
